### PR TITLE
Add support for the parameter-shift hessian to the TF interface

### PR DESCRIPTION
**Pull request by albi3ro**
This pull request ties in the parameter shift hessian calculation to the tensorflow interface.

Now you should be able to run:

```python
dev = qml.device('default.qubit', wires=1)
@qml.qnode(dev, interface='tf', diff_method='parameter-shift')
def circuit(x):
    qml.RX(x[0], wires=0)
    qml.RY(x[1], wires=0)
    return qml.expval(qml.PauliZ(0))

x = tf.Variable([0.1, 0.2], dtype=tf.float64)

with tf.GradientTape() as tape1:
    with tf.GradientTape() as tape2:
        y = circuit(x)
    grad = tape2.gradient(res, x)

tape1.jacobian(grad, x)
```

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,33 @@
 
 <h3>New features since last release</h3>
 
+- The TensorFlow interface supports calculating the Hessian. The Hessian
+  currently only works with `diff_method='parameter-shift`.
```
Comment (trbromley) [inline .github/CHANGELOG.md:6]: I asked a similar question for the torch PR, but here isn't there potential to support a backprop approach since we have `default.qubit.tf`?

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,33 @@
...
+          y = circuit(x)
+      grad = tape2.gradient(res, x)
+
+  hessian = tape1.jacobian(grad, x)
+  ```
+  You can also compute just the diagonal components by changing that last line to
+
+  ```python
+  hessian_diagonals = tape1.gradient(grad, x)
+  ```
```
Comment (trbromley) [inline .github/CHANGELOG.md:30]: Cool! Is it more efficient to do this?

File: tests/tape/test_jacobian_tape.py [hunk 3]
```diff
@@ -606,7 +606,7 @@ def test_non_differentiable_error(self):
             qml.RX(0.543, wires=[0])
             qml.RY(-0.654, wires=[1])
             qml.CNOT(wires=[0, 1])
-            qml.probs(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
```
Comment (trbromley) [inline tests/tape/test_jacobian_tape.py:609]: What was the reason for this change?

File: pennylane/interfaces/tf.py [hunk 4]
```diff
@@ -138,34 +138,74 @@ def _execute(self, params, **input_kwargs):
         res = self.execute_device(args, input_kwargs["device"])
         self.set_parameters(all_params, trainable_only=False)
 
-        def grad(grad_output, **tfkwargs):
-            variables = tfkwargs.get("variables", None)
+        saved_grad_matrices = {}
+
+        def _evaluate_grad_matrix(grad_matrix_fn):
```
Comment (trbromley) [inline pennylane/interfaces/tf.py:143]: Maybe worth adding the "Args:" part of the docstring to mention that `grad_matrix_fn` is a string

Could also mention that we expect it to be, e.g., "jacobian" or "hessian".

File: pennylane/interfaces/tf.py [hunk 5]
```diff
@@ -138,34 +138,74 @@ def _execute(self, params, **input_kwargs):
...
+
+            return grad_matrix
+
+        def jacobian_product(dy, **tfkwargs):
+            variables = tfkwargs.get("variables", None)
+            dy_row = tf.reshape(dy, [1, -1])
+
+            @tf.custom_gradient
+            def jacobian(p):
+                def hessian_product(ddy, **tfkwargs):
```
Comment (trbromley) [inline pennylane/interfaces/tf.py:180]: Not a problem for this PR, but this nested structure seems a not very maintainable way to provide higher-order gradients. I wonder why this is the standard approach.

File: tests/interfaces/test_qnode_tf.py [hunk 6]
```diff
@@ -522,6 +522,208 @@ def circuit():
...
+            [
+                [-0.5 * tf.cos(a) * tf.cos(b), 0.5 * tf.sin(a) * tf.sin(b)],
+                [0.5 * tf.sin(a) * tf.sin(b), -0.5 * tf.cos(a) * tf.cos(b)]
+            ],
+            [
+                [0.5 * tf.cos(a) * tf.cos(b), -0.5 * tf.sin(a) * tf.sin(b)],
+                [-0.5 * tf.sin(a) * tf.sin(b), 0.5 * tf.cos(a) * tf.cos(b)]
+            ]
+        ]
+        np.testing.assert_allclose(hess, expected_hess, atol=tol, rtol=0, verbose=True)
```
Comment (trbromley) [inline tests/interfaces/test_qnode_tf.py:725]: Nice, I didn't know about `np.testing.assert_allclose()`. Is there any advantage over `assert np.allclose()`? :thinking:

**Review by trbromley — APPROVED** [review]
Thanks @albi3ro! Really cool to see Hessian support coming in for the various interfaces :100:

File: .github/CHANGELOG.md (hunk 1 above)
Comment (josh146) [reply .github/CHANGELOG.md:6]: Yep, right - this line should be changed.

Comment (josh146) [inline .github/CHANGELOG.md:6]:
```suggestion
- The TensorFlow interface now supports computing second derivatives and Hessians of hybrid quantum models.
  Second derivatives are supported on both hardware and simulators.
```

File: .github/CHANGELOG.md (hunk 2 above)
Comment (josh146) [reply .github/CHANGELOG.md:30]: Not really no. Because PL currently always computes the full Jacobian/Hessian even when TF just wants the vjp/vhp :(

File: pennylane/interfaces/tf.py [hunk 7]
```diff
@@ -138,34 +138,74 @@ def _execute(self, params, **input_kwargs):
...
+            * Avoids duplicating logic surrounding parameter unwrapping/wrapping
+
+            * Takes advantage of closure, to cache computed gradient matrices via
+              saved_grad_matrices, to avoid gradient matrices being
+              computed multiple redundant times.
+
+              This is particularly useful when differentiating vector-valued QNodes.
+              Because TensorFlow requests the vector-GradMatrix product,
+              and *not* the full GradMatrix, differentiating vector-valued
+              functions will result in multiple backward passes.
```
Comment (josh146) [inline pennylane/interfaces/tf.py:158]:
```suggestion
              functions will result in multiple backward passes.

            Args:
                p (Sequence): quantum tape parameter values use to evaluate the gradient matrix
                grad_matrix_fn (str): Name of the gradient matrix function. Should correspond to an existing
                    tape method. Currently allowed values include ``"jacobian"`` and ``"hessian"``.

                Returns:
                    array[float]: the gradient matrix
```

File: pennylane/interfaces/tf.py (hunk 5 above)
Comment (josh146) [reply pennylane/interfaces/tf.py:188]: Because second derivatives and custom gradients are both not very common in current ML models 😆

File: tests/tape/test_jacobian_tape.py (hunk 3 above)
Comment (josh146) [reply tests/tape/test_jacobian_tape.py:609]: 🤔

File: pennylane/interfaces/tf.py [hunk 8]
```diff
@@ -138,34 +138,82 @@ def _execute(self, params, **input_kwargs):
...
-            variables = tfkwargs.get("variables", None)
+        saved_grad_matrices = {}
+
+        def _evaluate_grad_matrix(grad_matrix_fn):
+            """Convenience function for generating gradient matrices
+            for the given parameter values.
+
+            This function serves two purposes:
+
+            * Avoids duplicating logic surrounding parameter unwrapping/wrapping
```
Comment (thisac) [inline pennylane/interfaces/tf.py:149]:
```suggestion
            * Avoids duplicating logic surrounding parameter unwrapping/wrapping.

```

File: pennylane/interfaces/tf.py [hunk 9]
```diff
@@ -138,34 +138,82 @@ def _execute(self, params, **input_kwargs):
...
+
+            * Takes advantage of closure, to cache computed gradient matrices via
+              saved_grad_matrices, to avoid gradient matrices being
+              computed multiple redundant times.
+
+              This is particularly useful when differentiating vector-valued QNodes.
+              Because TensorFlow requests the vector-GradMatrix product,
+              and *not* the full GradMatrix, differentiating vector-valued
```
Comment (thisac) [inline pennylane/interfaces/tf.py:157]: Why is it written `GradMatrix` in camel case? I think it looks like a C++ class name like this. 🤔

File: tests/interfaces/test_qnode_tf.py (hunk 6 above)
Comment (thisac) [reply tests/interfaces/test_qnode_tf.py:728]: Also curious! 🤔

**Review by thisac — APPROVED** [review]
Looks nice @albi3ro! 👍

File: tests/interfaces/test_qnode_tf.py (hunk 6 above)
Comment (josh146) [reply tests/interfaces/test_qnode_tf.py:728]: `np.testing.assert_allclose()` provides a better traceback on failure, showing the full arrays it is trying to compare!

File: tests/tape/test_jacobian_tape.py (hunk 3 above)
Comment (josh146) [reply tests/tape/test_jacobian_tape.py:609]: @albi3ro was this change you or me?

File: pennylane/interfaces/tf.py (hunk 9 above)
Comment (josh146) [reply pennylane/interfaces/tf.py]: Updated to remove the camel case

**Review by josh146 — APPROVED** [review]
@albi3ro feel free to merge this one in!

File: tests/tape/test_jacobian_tape.py (hunk 3 above)
Comment (albi3ro) [reply tests/tape/test_jacobian_tape.py:609]: I changed it back when we couldn't support calculate the hessian of a vector-valued output.  We can now take the hessian of a vector valued function, so I'm changing it back.

Comment (josh146) [reply tests/tape/test_jacobian_tape.py]: @albi3ro in retrospect this was actually a good change, as it increased coverage! I think if you change it back again (sorry 😬 ) codecov will pass and this can be merged in)
